lib/workflows/send_decom_image_email.py

```diff
@@ -0,0 +1,282 @@
+    use_override: bool = False,
+    override_email_address: Optional[str] = "cloud-support@stfc.ac.uk",
+    cc_cloud_support: bool = False,
+    **email_params_kwargs,
+):
+    """
+    Sends an email to each user who owns one or more VMs that are running an image that is to be decommissioned.
+    This email will contain a notice to delete or rebuild the VM
+    :param smtp_account: (SMTPAccount): SMTP config
+    :param cloud_account: string represents cloud account to use
+    :param image_name_list: A list of image names to be decommissioned
+    :param image_eol_list: A list of End of Life (EOL) date strings (YY/MM/DD) for images to be decommissioned
+    :param image_upgrade_list: A list of image names that users are recommended to upgrade to
+    :param limit_by_projects: A list of project names or ids to limit search in
+    :param all_projects: A boolean which if selected will search in all projects
+    :param send_email: Actually send the email instead of printing what will be sent
+    :param as_html: Send email as html
+    :param use_override: flag if set will use override email address
+    :param override_email_address: an overriding email address to use if override_email set
+    :param cc_cloud_support: flag if set will cc cloud-support email address to each generated email
+    :param email_params_kwargs: see EmailParams dataclass class docstring
+    """
+    assert (limit_by_projects or all_projects) and not (
+        limit_by_projects and all_projects
+    ), "Choose either a list of project identifiers or use the 'all_projects' flag"
+
+    decom_image_info = get_image_info(
+        image_name_list, image_eol_list, image_upgrade_list
+    )
+    server_query = find_servers_with_decom_images(
+        cloud_account, image_name_list, limit_by_projects
+    )
+
+    for user_id in server_query.to_props().keys():
+        # if email_address not found - send to override_email_address
+        # also send to override_email_address if override_email set
+        user_name, email_addr = find_user_info(
+            user_id, cloud_account, override_email_address
+        )
+        send_to = [email_addr]
+        if use_override:
+            send_to = [override_email_address]
+
+        if not send_email:
+            print_email_params(
+                send_to[0],
+                user_name,
+                as_html,
+                get_affected_images_plaintext(decom_image_info),
+                server_query.to_string(groups=[user_id], include_group_titles=False),
+            )
+
+        else:
+            email_params = build_email_params(
+                user_name,
+                get_affected_images_plaintext(decom_image_info)
+                if not as_html
+                else get_affected_images_html(decom_image_info),
+                server_query.to_string(groups=[user_id])
+                if not as_html
+                else server_query.to_html(groups=[user_id], include_group_titles=False),
```

Comment:
Nit: can we do these if-else outside of this method call as it's a bit difficult to read